### ISSUE TITLE
Lot of ram consumption

### ISSUE THREAD

**Issue by wiiznokes** [elevenhsoft/cosmic-ext-applet-ollama#5]
![screenshot-2024-06-15-15-47-11](https://github.com/elevenhsoft/cosmic-ext-applet-ollama/assets/78230769/c22bcbd1-0b55-47f4-bd22-5e4a66e8025a)

I think there might be a memory leak somewhere. The memory usage seems to keep going up.

**Comment by elevenhsoft** [comment]
Good find! It might be because we using `Box::leak()` in markdown widget. 
We should find good solution for this. There are many calls for this widget from bot response so I assume this is the issue.
If we will get rid of Boxed strings, it should be good.

**Comment by elevenhsoft** [comment]
I think I sorted it out, please take a look: 
![Screenshot_20240615_210151](https://github.com/elevenhsoft/cosmic-ext-applet-ollama/assets/114903054/c9233d46-0e48-40aa-a42c-27f33b29c05f)

80MB ram in debug mode, with pretty long response.

**Comment by elevenhsoft** [comment]
@wiiznokes I made some changes. Would you like to test it again? Would be nice :)

**Comment by wiiznokes** [comment]
I'm building... :smiley:

**Comment by wiiznokes** [comment]
Nop, still taking lot of ram on my end /:

**Comment by elevenhsoft** [comment]
Damn... I have no idea what else can cause it. On my end everything is fine, without any leaks. Spotify takes more RAM than my applet :P

I though it may be related to some optimizations while building for release but no, just installed it and still nothing.

**Comment by wiiznokes** [comment]
But how is it possible that you don't have any leaks ? From what i see, there is still the `Box::leak` in the `Markdown::new` function, which is in a very hot path, since it is called in the view function.

From what i see, cosmic-edit store the Editor struct in app state: https://github.com/pop-os/cosmic-edit/blob/454a68f05bfb1783b8b196a8e566c406c076fd9a/src/tab.rs#L38.

So, imo, you should have a vec of `Editor` in app state. If `Editor` is not able to own the spans, you can always self ref using https://github.com/someguynamedjosh/ouroboros

**Comment by elevenhsoft** [comment]
> But how is it possible that you don't have any leaks ? From what i see, there is still the `Box::leak` in the `Markdown::new` function, which is in a very hot path, since it is called in the view function.

Yeah but it should not be called many times from what I understand. It's not a rendering method.



> From what i see, cosmic-edit store the Editor struct in app state: https://github.com/pop-os/cosmic-edit/blob/454a68f05bfb1783b8b196a8e566c406c076fd9a/src/tab.rs#L38.
> 
> So, imo, you should have a vec of `Editor` in app state. If `Editor` is not able to own the spans, you can always self ref using https://github.com/someguynamedjosh/ouroboros

Thank you. Tomorrow I will try to solve it somehow. Maybe I will find a better approach based on this :)

**Comment by wiiznokes** [comment]
> Yeah but it should not be called many times from what I understand. It's not a rendering method.

Afaik, every time the mouse is mooving around the windows for example. You should try to print a debug message to see.

**Comment by elevenhsoft** [comment]
Yes but it's not from `new()` method but `draw()`. I did not try to check it on `new()` however. 
I think it should be safe but definitely it needs improvement. I will try to fix it because it's not a best solution right now anyway.

@mmstick said will try to fix those leaks soon so maybe it will help.

**Comment by elevenhsoft** [comment]
Ok, just checked. indeed, there are few calls from `new()` method. Not that much but it's definitely calling few times in some time. Maybe moving Editor to state will be good idea. 

Let me think tomorrow :)
